group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/GeneralUniformAssignmentBuilder.java

```diff
@@ -14,17 +14,903 @@
+            int maxAssignmentSize = assignmentManager.maxAssignmentSize(member);
+            if (memberPartitionCount == maxAssignmentSize)
+                continue;
+
+            // Otherwise make sure it cannot get any more partitions.
+            for (Uuid topicId : members.get(member).subscribedTopicIds()) {
+                Set<Integer> assignedPartitions = targetAssignment.get(member).targetPartitions().get(topicId);
+                for (int i = 0; i < subscribedTopicDescriber.numPartitions(topicId); i++) {
+                    TopicIdPartition topicIdPartition = new TopicIdPartition(topicId, i);
+                    if (assignedPartitions == null || !assignedPartitions.contains(i)) {
+                        String otherMember = partitionOwnerInTargetAssignment.get(topicIdPartition);
+                        int otherMemberPartitionCount = assignmentManager.targetAssignmentSize(otherMember);
+                        if (memberPartitionCount + 1 < otherMemberPartitionCount) {
+                            LOG.debug("{} can be moved from member {} to member {} for a more balanced assignment.",
+                                topicIdPartition, otherMember, member);
+                            return false;
+                        }
+                    }
+                }
+            }
+        }
+        return true;
+    }
+
+    /**
+     * Balance the current assignment after the initial round of assignments have completed.
+     */
+    private void balance() {
+        if (!unassignedPartitions.isEmpty())
+            throw new PartitionAssignorException("Some partitions were left unassigned");
+        // Refill unassigned partitions with all the topicId partitions.
+        unassignedPartitions.addAll(topicIdPartitions(subscribedTopicIds, subscribedTopicDescriber));
+
+        // Narrow down the reassignment scope to only those partitions that can actually be reassigned.
+        Set<TopicIdPartition> fixedPartitions = new HashSet<>();
+        for (Uuid topicId : subscribedTopicIds) {
+            if (!canTopicParticipateInReassignment(topicId)) {
+                for (int i = 0; i < subscribedTopicDescriber.numPartitions(topicId); i++) {
+                    fixedPartitions.add(new TopicIdPartition(topicId, i));
+                }
+            }
+        }
+        unassignedPartitions.removeAll(fixedPartitions);
+
+        // Narrow down the reassignment scope to only those members that are subject to reassignment.
+        for (String member : members.keySet()) {
+            if (!canMemberParticipateInReassignment(member)) {
+                sortedMembersByAssignmentSize.remove(member);
+            }
+        }
+
+        // If all the partitions are fixed i.e. unassigned partitions is empty there is no point of re-balancing.
+        if (!unassignedPartitions.isEmpty()) performReassignments();
+    }
+
+    private void performReassignments() {
```

Comment:
I am a bit worried about the rebalance part, especially with regarding to how it will scale with the number of member and partitions in the group. Should we file a jira to look into this a bit more closely at some point? It seems that there are opportunities to make it perform better. For instance, we could try to make it incremental.

Reply:
hey I was worried about this too, I've thought of multiple ways to make it better, hence it took so long to implement, but seems like I've run out of ideas and any idea I had came with its own problems. I'm open to discussing it again.

Reply:
Could you please elaborate more on making it incremental?

Reply:
Created https://issues.apache.org/jira/browse/KAFKA-14048.